tst/unit/kokkos_abstraction.cpp

```diff
@@ -262,6 +262,175 @@ TEST_CASE("par_for loops", "[wrapper]") {
   }
 }
 
+template <class OuterLoopPattern, class InnerLoopPattern>
+bool test_wrapper_nested_3d(OuterLoopPattern outer_loop_pattern,
+                            InnerLoopPattern inner_loop_pattern,
+                            DevExecSpace exec_space) {
+  // Compute the 2nd order centered derivative in x of i+1^2 * j+1^2 * k+1^2
+
+  const int N = 32;
+  ParArray3D<Real> dev_u("device u", N, N, N);
+  ParArray3D<Real> dev_du("device du", N, N, N - 2);
+  auto host_u = Kokkos::create_mirror(dev_u);
+  auto host_du = Kokkos::create_mirror(dev_du);
+
+  // initialize with i^2 * j^2 * k^2
+  for (int n = 0; n < N; n++)
+    for (int k = 0; k < N; k++)
+      for (int j = 0; j < N; j++)
+        for (int i = 0; i < N; i++)
+          host_u(k, j, i) = pow((i + 1) * (j + 1) * (k + 1), 2.0);
+
+  // Copy host array content to device
+  Kokkos::deep_copy(dev_u, host_u);
+
+  // Compute the scratch memory needs
+  const int scratch_level = 0;
+  size_t scratch_size_in_bytes = parthenon::ScratchPad1D<Real>::shmem_size(N);
+
+  // Compute the 2nd order centered derivative in x
+  parthenon::par_outer_for(
```

Comment:
For some reason, I feel like `par_for_outer` is more natural than `par_outer_for`. But of course, that's just personal preference. The name doesn't need to change.

Reply:
> For some reason, I feel like `par_for_outer` is more natural than `par_outer_for`. 

I realize that `par_for_outer` more sense from a logical grouping with `par_for`, I only used `par_outer_for` because it made more sense when spoken. I'll change to `par_for_outer` and `par_for_inner` though